feat: P4ADEV-3827 draft of the service and registry event type

dan-ntt opened the pull request:
#### Description
<!--- Please always add a PR description as if nobody knows anything about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the same page. -->
<!--- Write this PR as you were contributing to a public OSS project, where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->

#### List of Changes
<!--- Describe your changes in detail -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- Pre-Deploy Test
  - [x] Unit
  - [ ] Integration (Narrow)
- Post-Deploy Test
  - [ ] Isolated Microservice
  - [ ] Broader Integration
  - [ ] Acceptance
  - [ ] Performance & Load

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] PATCH - Bug fix (backwards compatible bug fixes)
- [x] MINOR - New feature (add functionality in a backwards compatible manner)
- [ ] MAJOR - Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] CHORE - Minor Change (fix or feature that don't impact the functionality e.g. Documentation or lint configuration)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Commit:
feat: draft of the service and registry event type

Commit:
feat: changes and implementation of the new refactored service

Commit:
fix: fixed the failing tests

Commit:
feat: set to return empty on client call because the feat on debt position is not merged

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/connector/debtpositions/client/DebtPositionClient.java:
```diff
@@ -69,4 +70,9 @@ public DebtPosition getDebtPositionByInstallmentId(Long installmentId, String ac
     }
   }
+
+  public List<DebtPositionDTO> getDebtPositionsByIdentificativoUnivocoPersonaFGAndOrganizationId(CtIdentificativoUnivocoPersonaFG identificativoUnivocoPersonaFG, Long organizationId, String accessToken) {
```
can we use PU names instead of `IdentificativoUnivocoPersonaFG`? let's try to hide this kind of name on the presentation layer (or the service directly called)

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediPosizioniAperteService.java:
```diff
@@ -0,0 +1,110 @@
+            PaaSILPosizioniAperte openPosition = new PaaSILPosizioniAperte();
+            openPosition.setCodIpaEnte(request.getCodIpaEnte());
+            openPosition.setDeNomeEnte(organization.getOrgName());
+            openPosition.setUrlPagamento(null); // TODO: to be defined
```
can we resolve this?

antonioT90 replied:
@robertopasetti, to resolve this TODO we have this hint from @serdimic 

> questa parte non è direttamente migrabile da MyPay a PU, per via di differenze nell'architettura; l'idea potrebbe essere quella di comporre una url di una API di PU che ha nel path o nella query un qualche identificativo del dovuto (potrebbe essere l'installmentId, o qualcosa del genere). Invocata, tale url a sua volta invoca le API ci checkout del nodo e ritorna la url di checkout che innesca il pagamento (c'è qualcosa di simile sull'api paaSILVerificaAvviso). Il problema è che questa API di PU dovrebbe essere pubblica e non sottoposta a token JWT, visto che sarebbe invocata direttamente dal cittadino. Forse conviene coinvolgere Antonio Torre e Roberto Pasetti su questo punto.

could we use some URL exposed by `ARpu` to resolve it?

robertopasetti replied:
we should call this API https://developer.pagopa.it/pago-pa/guides/sanp/appendici/primitive#ec-checkout-api that returns an URL that can be inovoked by another application. This should be the case:

- Broker App calls PU WS paaSILChiediPosizioniAperte
- paaSILChiediPosizioniAperte find every unpaid installment
- for each unpaid installment PU calls PagoPA API Carts (use returnOkUrl, returnCancelUrl, returnErrorUrl used in another WS cc @serdimic  )
- PagoPA API Carts returns an URL
- paaSILChiediPosizioniAperte returns that Url to Broker App

robertopasetti replied:
the URL returned from PagoPA API Carts is public

robertopasetti replied:
inviaDovuti should returns a PagoPA checkout's URL, isnt'it?

serdimic replied:
but consider that this API returns all open positions for a citized. This means that if a citizen has for instance 10 open position, you would need to invoke 10 times checkout API in a synchronous way.
I don't think it's the best way to return the ckeckout url in the paaSILChiediPosizioniAperte, it would be an overkill.
It's just like in the "area riservata" you would geenrate and return the checkout URL for every open position searched by a citizen.

serdimic replied:
> inviaDovuti should returns a PagoPA checkout's URL, isnt'it?

paaSILInviaDovuti is used in a quite different scenario: the SIL asks explicitely PU to generate a single position and provide a url for the citizen in order to pay it (that is the main reason the position is created). So it's reasonable to return the checkout url.

paaSILChiediPosizioniAperte is mainly used to provide citized with a list of his open position, not on a specific organization but for every organization managed by PU. The payment url it's just provided to allow citizen - only if he wants - to start a payment for one of the found open position. That's the reason I would consider an overkill to trigger a cart creation on Checkout API for every open position processed by paaSILChiediPosizioniAperte.

robertopasetti replied:
and what happened before in MyPay? It returned an URL for every open position?

robertopasetti replied:
ok, it's clear. We can improve this implementation in the next sprints. We should expose an unauthenticated URL from PU (this is what this WS will return as UrlPagamento) that called from Broker App redirects to checkout.

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediPosizioniAperteService.java:
```diff
@@ -0,0 +1,110 @@
+      .toList();
+  }
+
+  private Dovuti buildDovuti(PaymentOptionDTO paymentOption, InstallmentDTO installment) {
```
@serdimic there is something we could re-use for this mapping?

serdimic replied:
I would say no, in the other WS of PU-SIL where the object `Dovuti` is used, it's only unmarshalled since it is just used to retrieve data from the SOAP request. Here on contrary it's created and provided (encoded) in the response.
Just as a reference, the unmarashalling currently occurs in classes `PaaSILInviaCarrelloDovutiMapper` and `PaaSILInviaDovutiMapper` of package `it.gov.pagopa.pu.sil.mapper`.

Commit:
fix: implemented the changes to API handler

Commit:
test: implemented tests for open positions

Commit:
feat: changes based on review

Commit:
fix: fixed linting issues

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/connector/debtpositions/DebtPositionService.java:
```diff
@@ -68,4 +69,22 @@ public interface DebtPositionService {
+   * @param identificativoUnivocoPersonaFG the CtIdentificativoUnivocoPersonaFG object
+   * @param organizationId the ID of the organization
+   * @param accessToken the access token for authentication
+   * @return a list of DebtPositionDTOs matching the criteria
```
the doc is referencing old names

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediPosizioniAperteService.java:
```diff
@@ -0,0 +1,156 @@
+    CtDatiSingoloVersamentoDovuti datiSingoloVersamento = of.createCtDatiSingoloVersamentoDovuti();
+    datiSingoloVersamento.setIdentificativoUnivocoDovuto(installment.getIud());
+    datiSingoloVersamento.setImportoSingoloVersamento(ConversionUtils.centsAmountToBigDecimalEuroAmount(installment.getAmountCents()));
+    datiSingoloVersamento.setCommissioneCaricoPA(ConversionUtils.centsAmountToBigDecimalEuroAmount(installment.getNotificationFeeCents()));
```
@serdimic @robertopasetti is this right?

robertopasetti replied:
no, it's wrong. we talked about it yesterday with @dan-ntt . At this moment we don't manage 'CommissioneCaricoPA' in PU. By the way @serdimic what's its meaning ?

serdimic replied:
It's just an informative field of "dovuto" entity (i.e. installment), that may optionally be set by the creator of "dovuto" (SIL by WS, operator by GUI).
In MyPay, to my knowledge, it has no role in any process / business logic, and its value is never processed by MyPay; it's just stored exactly as passed, and its value can be retrieved on UI and in the export files.

Don't have any detailed description of its functional meaning, the only clue is its name.

antonioT90 replied:
@serdimic of we set it to null? if the SIL should set it when its creating it through WS, actually on PU are storing it or ignoring it?

serdimic replied:
we are just ignoring it in PU (for instance, see `it.gov.pagopa.pu.sil.mapper.PaaSILImportaDovutoMapper#fillInstallmentFields`), since for what I know there is no corresponding field in PU entities (in particular Installment).

antonioT90 replied:
ok.. @robertopasetti do you think we should add this field and storing it just to return it...
Or we should ignore it and return null?

robertopasetti replied:
ignore it for the moment

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediPosizioniAperteService.java:
```diff
@@ -0,0 +1,156 @@
+            openPosition.setCodIpaEnte(request.getCodIpaEnte());
+            openPosition.setDeNomeEnte(organization.getOrgName());
+
+            try {
+              CartRequest cartRequest = cartRequestMapper.mapInstallmentToCartRequest(installment, organization, null, null);
+              openPosition.setUrlPagamento(checkoutClient.checkoutCart(cartRequest));
+            } catch (Exception e) {
+              log.error("Error generating payment URL for installment[{}]", installment.getInstallmentId(), e);
+            }
```
as discussed before, this URL should not be checkout URL...
@robertopasetti here we should set a TODO to remember to set ARpu URL instead?

see: https://github.com/pagopa/p4pa-pu-sil/pull/158/files/de5de34038537692d64b86f18eae87a7f2ce6ec5#r2416044815

robertopasetti replied:
URLPagamento is the checkout URL in order to finalize the payment
URLDownloadRT is the URL where ARPU should expose a public URL in order to show the receipt PDF

antonioT90 replied:
@robertopasetti yes, but if you read the discussion linked, @serdimic was proposing to not configure checkout because it could be overwhelming

dan-ntt replied:
@antonioT90 we decided that for now we use the checkout and later on update it with a new one

robertopasetti replied:
yes, I remember. I will open a tasks to analyze this issue. For the moment let's keep checkout's url

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediStoricoPagamentiService.java:
```diff
@@ -0,0 +1,107 @@
+      throw new SilFaultException(SilFaults.PAA_ENTE_NON_VALIDO, "L'ente non è valido o non è abilitato");
+    }
+
+    OffsetDateTime dateFrom = LocalDate.now().atStartOfDay().atOffset(ZoneId.systemDefault().getRules().getOffset(Instant.now()));
```
```suggestion
    OffsetDateTime dateFrom = OffsetDateTime.now().truncatedTo(ChronoUnit.DAYS);
```

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediStoricoPagamentiService.java:
```diff
@@ -0,0 +1,107 @@
+      dateFrom = request.getDataFrom().toGregorianCalendar().toZonedDateTime().toOffsetDateTime();
+    }
+
+    OffsetDateTime dateTo = LocalDate.now().atTime(LocalTime.MAX).atOffset(ZoneId.systemDefault().getRules().getOffset(Instant.now()));
```
you could store the dateFrom default value (currentDateStart) and here just call currentDateStart.plusDays(1)

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediStoricoPagamentiService.java:
```diff
@@ -0,0 +1,107 @@
+
+    OffsetDateTime dateFrom = LocalDate.now().atStartOfDay().atOffset(ZoneId.systemDefault().getRules().getOffset(Instant.now()));
+    if (request.getDataFrom() != null) {
+      dateFrom = request.getDataFrom().toGregorianCalendar().toZonedDateTime().toOffsetDateTime();
```
can we use `it.gov.pagopa.pu.sil.util.ConversionUtils#toOffsetDateTime(javax.xml.datatype.XMLGregorianCalendar)` instead?

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediStoricoPagamentiService.java:
```diff
@@ -0,0 +1,107 @@
+
+    OffsetDateTime dateTo = LocalDate.now().atTime(LocalTime.MAX).atOffset(ZoneId.systemDefault().getRules().getOffset(Instant.now()));
+    if (request.getDataTo() != null) {
+      dateTo = request.getDataTo().toGregorianCalendar().toZonedDateTime().toOffsetDateTime();
```
can we use `it.gov.pagopa.pu.sil.util.ConversionUtils#toOffsetDateTime(javax.xml.datatype.XMLGregorianCalendar)` instead?

Commit:
fix: implemented changes based on PR comments
- used `ConversionUtils` for OffsetDateTime transformation from XMLGregorianCalendar
- added todo for payment URL
- simplified the generation of dateFrom and dateTo (defaults)

Commit:
fix: updated the test for endpoints

Commit:
fix: removed set commissions

antonioT90 commented on src/main/java/it/gov/pagopa/pu/sil/service/querypayments/PaaSILChiediPosizioniAperteService.java:
```diff
@@ -0,0 +1,155 @@
+
+            try {
+              CartRequest cartRequest = cartRequestMapper.mapInstallmentToCartRequest(installment, organization, null, null);
+              openPosition.setUrlPagamento(checkoutClient.checkoutCart(cartRequest)); // @TODO: to replace with a different URL
```
There is a JIRA ticket to track this TODO? If yes, let's write It here

robertopasetti replied:
here it is https://pagopa.atlassian.net/browse/P4ADEV-4043

Commit:
fix: added ticket to TODO
